### ISSUE TITLE
Delays for non-JPEG mode caused by GC at regular intervals

### ISSUE THREAD

**Issue by samhed** [kanaka/noVNC#559]
We have been looking at the performance of the new zlib code among other things when we noticed this new behavior.

I disabled JPEG by running the following in the web console:

    UI.rfb._encodings.splice(8,1)
    RFB.messages.clientEncodings(UI.rfb._sock, UI.rfb._encodings, true, true)

The garbage collector kicks in consistently at an interval of ~7 seconds and this causes a delay of about 50-150ms each time.

The old code only has these incidents sporadically. I have been trying to figure out why, but I cant find what is causing this.. Any ideas @DirectXMan12 ?

**Comment by samhed** [comment]
The graph in Firefox's performance measurement tool illustrates this quite well.

![screenshot from 2015-12-04 16 39 16](https://cloud.githubusercontent.com/assets/4117162/11593843/d116f23e-9aa6-11e5-9f2f-518357cbee14.png)

Playing a video in the session is a good way to make the delays noticeable.

**Comment by DirectXMan12** [comment]
hmm... I'll try and take a look next week.  This only happens when JPEG is disabled?  It's possible the server is sending one of the other encodings instead, and there's a bug in one of those.

**Comment by samhed** [comment]
> This only happens when JPEG is disabled?

yes

**Comment by DirectXMan12** [comment]
I think I've been able to reproduce what you're seeing (using TigerVNC) and I've identified a couple of potential improvements.  I still get GC pauses, but they're fewer and farther between.  I'll try and post a branch with the fixes soon so that you can test them out and see if things improve for you as well.

**Comment by samhed** [comment]
that sounds great!

**Comment by DirectXMan12** [comment]
Sorry about the delay -- I've posted a couple of changes which seem to reduce GC pauses for me.  They need some tweaking before being production-ready, but I wanted to see if they help you out at all: kanaka/novnc#566.  PTAL and let me know if they help at all.

**Comment by samhed** [comment]
In fact I can't reproduce these problems which I had in the first place. According to the yum logs on the machine on which I found this, it had Firefox 38 at the time, which also seems to be correct judging from the look of the "performance"-tool in the screenshot. But with the old code I can't reproduce the issue with Firefox 37. 38, 40, 41, 42, 43 or 44. Firefox 39 crashed when I tried to measure the performance.

I don't have access anymore to the specific machine on which I found this problem originally. Perhaps different graphics cards or an update to some library which Firefox uses is the cause of this. I don't know.

It's quite difficult to say if the issue is fixed if I can't reproduce it again in the first place.
